src/Model_Hydro/MHD_Init_BField_ByVecPot_Function.cpp

```diff
@@ -0,0 +1,187 @@
+
+
+
+//-------------------------------------------------------------------------------------------------------
+// Function    :  MHD_Init_BField_ByVecPot_Function
+// Description :  Use the function pointer "Init_BField_ByVecPot_User_Ptr"
+//                to assign a magnetic vector potential to all real patches on level
+//                "B_lv" and take the curl of this potential to compute the magnetic field
+//
+// Note        :
+//
+// Parameter   :  B_lv : Target AMR level
+//
+// Return      :  amr->patch->magnetic
+//-------------------------------------------------------------------------------------------------------
+void MHD_Init_BField_ByVecPot_Function( const int B_lv )
+{
+
+// check
+#  ifndef MHD
+   Aux_Error( ERROR_INFO, "MHD must be enabled !!\n" );
+#  endif
+
+   if ( Init_BField_ByVecPot_User_Ptr == NULL  &&  OPT__INIT_BFIELD_BYVECPOT == 2 )
+      Aux_Error( ERROR_INFO, "Init_BField_ByVecPot_User_Ptr == NULL !!\n" );
+
+
+   if ( MPI_Rank == 0 )   Aux_Message( stdout, "   Constructing the magnetic field from vector potential ...\n" );
+
+
+// set the number of OpenMP threads
+#  ifdef OPENMP
+   const int OMP_NThread = ( OPT__INIT_GRID_WITH_OMP ) ? OMP_NTHREAD : 1;
+#  else
+   const int OMP_NThread = 1;
+#  endif
+
+
+// allocate memory for per-thread arrays
+   double **Ax = NULL, **Ay = NULL, **Az = NULL;
+
+   Aux_AllocateArray2D( Ax, OMP_NThread, CUBE(PS1+1) );
+   Aux_AllocateArray2D( Ay, OMP_NThread, CUBE(PS1+1) );
+   Aux_AllocateArray2D( Az, OMP_NThread, CUBE(PS1+1) );
+
+
+// loop over the patches on the target AMR level
+   const double dh = amr->dh[B_lv];
+
+#  pragma omp parallel for schedule( runtime ) num_threads( OMP_NThread )
+   for (int PID=0; PID<amr->NPatchComma[B_lv][1]; PID++)
+   {
+
+#     ifdef OPENMP
+      const int TID = omp_get_thread_num();
+#     else
+      const int TID = 0;
+#     endif
+
+      const double EdgeL[3] = { amr->patch[0][B_lv][PID]->EdgeL[0],
+                                amr->patch[0][B_lv][PID]->EdgeL[1],
+                                amr->patch[0][B_lv][PID]->EdgeL[2] };
+
+
+//    set the magnetic vector potential from function
+      for (int k=0; k<PS1+1; k++) {  const double z0 = EdgeL[2] + k*dh;  const double zc = z0 + 0.5*dh;
+      for (int j=0; j<PS1+1; j++) {  const double y0 = EdgeL[1] + j*dh;  const double yc = y0 + 0.5*dh;
+      for (int i=0; i<PS1+1; i++) {  const double x0 = EdgeL[0] + i*dh;  const double xc = x0 + 0.5*dh;
+
+         int idx = IDX321( i, j, k, PS1+1, PS1+1 );
+
+         Ax[TID][idx] = 0.0;
+         Ay[TID][idx] = 0.0;
+         Az[TID][idx] = 0.0;
+
+         if ( i != PS1 )   Init_BField_ByVecPot_User_Ptr( &Ax[TID][idx], xc, y0, z0, Time[B_lv], B_lv, 'x', NULL );
+         if ( j != PS1 )   Init_BField_ByVecPot_User_Ptr( &Ay[TID][idx], x0, yc, z0, Time[B_lv], B_lv, 'y', NULL );
+         if ( k != PS1 )   Init_BField_ByVecPot_User_Ptr( &Az[TID][idx], x0, y0, zc, Time[B_lv], B_lv, 'z', NULL );
+
+      }}}
```

Comment:
I will fix the incorrect algorithm later.

On the implementation, I prefer to separate the features in another function (`MHD_Init_BField_ByVecPot_Function()` or `MHD_Init_BField_ByVecPot_Anal()`. The main reasons are
- Not need to disable the code for loading data from `B_IC` in `MHD_Init_BField_ByVecPot_File()` when `OPT__INIT_BFIELD_BYVECPOT != 1`. In fact, it is not difficult to update `MHD_Init_BField_ByVecPot_File()`, but I think it could make the code less readable.
- We can support OpenMP parallelism (`OPT__INIT_GRID_WITH_OMP`) for initializing the B field from vector potential, where I think it is difficult to be implemented in `MHD_Init_BField_ByVecPot_File()` due the HDF5 routine within the main loop.

What do you think?

Reply:
@hfhsieh I agree